### DISCUSSION TITLE
Render only a single component entirely on the client? | Monaco editor component

### DISCUSSION THREAD

**Discussion by dikkadev** [leptos-rs/leptos#1194] [Q&A] [answered]
I want to make a Monaco component with the [rust-monaco crate](https://github.com/siku2/rust-monaco).

For this, the component needs to be rendered on the client as the server, of course, cannot use the bindings. Is there a way I can tell Leptos to not SSR a single component or force it to re-render it on the client?

I could also be on the wrong track about how this needs to work, so if anyone has already made monaco component for Leptos, I'd appreciate the help.

**Comment by gbj** [comment]
Hm this is interesting. It's definitely somewhere on the agenda to have something like a `<ClientOnly/>` component.

For now, the closest you can get is probably
```rust
let (outlet, set_outlet) = create_signal(cx, None::<View>);
// `create_effect` only runs on the client, and will run once immediately
create_effect(cx, move |_| {
  // delays a tick until after hydration is done so there's no mismatch
  request_animation_frame(move || {
    set_outlet(/* whatever the new view is */)
  })
})
```
This could be wrapped in a component, of course. In fact, that's probably pretty darn close to what a `<ClientOnly/>` component would look like...

**Reply by dikkadev** [reply]
I'm not sure if this should still be in this Discussion, if not I can open a new one. But with this the new view is only partly set/updated.

```rust
let (outlet, set_outlet) = create_signal(cx, None::<HtmlElement<Div>>);
    // `create_effect` only runs on the client, and will run once immediately
    create_effect(cx, move |_| {
        cfg_if::cfg_if! {
            if #[cfg(feature = "hydrate")] { //compile this only for wasm code
                //...
                // delays a tick until after hydration is done so there's no mismatch
                request_animation_frame(move || {
                    let div = view! {cx, <div class="w-11/12 h-full border"></div>};
                    //... a &div is passed to the monaco create function
                    l!("div: {:?}", div.outer_html()); //shows the expected html of the whole editor

                    /*
                    Whats in the DOM:
                    <div class="w-11/12 h-full border" style="--codelens-font-features_ee1f61: "liga" off, "calt" off;">
                    </div>
                     */

                    set_outlet(Some(div));
                })
            }
        }
    });
    view! {cx,
        {move || match outlet.get() {
            Some(v) => v,
            None => view! {cx, <div>"loading or something"</div>}
        }}
    }
```
Do I need to access it differently? Still quite new to Leptos, there's a lot to wrap your head around and keep in mind sometimes.

_The [create function](https://docs.rs/monaco/latest/monaco/api/struct.CodeEditor.html#method.create) mentioned in the comment_

**Reply by gbj** [reply, marked as answer]
1) I gave you a really dumb way to do this. Here's one that's much better:
```rust
    let editor_ref = create_node_ref::<html::Div>(cx);
    create_effect(cx, move |_| {
        use wasm_bindgen::JsCast;

        if let Some(div) = editor_ref.get() {
            let editor = monaco::api::CodeEditor::create(div.unchecked_ref(), Some(monaco::api::CodeEditorOptions::default()));
            on_cleanup(cx, move || drop(editor));
            log!("div: {:?}", div.outer_html());
        }
    });
    view! { cx,
        <div class="w-11/12 h-full border" node_ref=editor_ref></div>
    }
```
2) The reason for the behavior, which I figured out after looking at the `rust-monaco` docs, is that they implement `Drop` for the `CodeEditor` type such that it [removes the code editor from the DOM](https://docs.rs/monaco/0.3.0/src/monaco/api/editor.rs.html#238-242). So I stuck it in an `on_cleanup` here so that it won't be cleaned up until the element is removed.

**Reply by dikkadev** [reply]
This works perfectly! I also tried to somehow do it with node_ref, but couldn't figure it out.

Appreciate the help, and the explanation!
